x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java

```diff
@@ -0,0 +1,189 @@
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+import org.elasticsearch.xpack.ql.expression.function.TwoOptionalArguments;
+import org.elasticsearch.xpack.ql.expression.function.aggregate.EnclosedAgg;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.sql.type.SqlDataTypeConverter;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.toMap;
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isFoldable;
+
+public abstract class HasPercentilesConfig extends NumericAggregate implements EnclosedAgg, TwoOptionalArguments {
+
+    // extensive list of all the possible PercentilesMethods (all can be used with default method parameters)
+    private static final Map<String, PercentilesMethod> NAME_TO_METHOD = stream(PercentilesMethod.values()).collect(toMap(
+        pm -> pm.getParseField().getPreferredName(),
+        pm -> pm
+    ));
+
+    // list of all the possible PercentileMethods that can we are capable of parameterizing as of now
+    private static final Map<PercentilesMethod, MethodParameterResolver> METHOD_TO_RESOLVER = Map.of(
+        PercentilesMethod.TDIGEST, TypeResolutions::isNumeric,
+        PercentilesMethod.HDR, TypeResolutions::isInteger
+    );
+    
+    private final Expression method;
+    private final Expression methodParameter;
+    
+    public HasPercentilesConfig(Source source, Expression field, List<Expression> parameters, 
+        Expression method, Expression methodParameter) 
+    {
+        super(source, field, parameters);
+        this.method = method;
+        this.methodParameter = methodParameter;
+    }
+
+    @Override 
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = super.resolveType();
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return resolvePercentileConfiguration(sourceText(), method, Expressions.ParamOrdinal.fromIndex(parameters().size()+1), 
+            methodParameter, Expressions.ParamOrdinal.fromIndex(parameters().size()+2));
+    }
+
+    public Expression method() {
+        return method;
+    }
+
+    public Expression methodParameter() {
+        return methodParameter;
+    }
+
+    public PercentilesConfig percentilesConfig() {
+        return asPercentilesConfig(method, methodParameter);
+    }
+
+    @FunctionalInterface
+    interface MethodParameterResolver {
+        Expression.TypeResolution resolve(Expression methodParameter, String sourceText, Expressions.ParamOrdinal methodParameterOrdinal);
+    }
+
+    private static Expression.TypeResolution resolvePercentileConfiguration(
```

Comment:
There's no value in making this method static - in fact it hurts. The method/params should be taken from the enclosing class simplifying the signature.
The `NAME_TO_METHOD` (has little meaning to say `ENUM_RESOLUTION`) can contain `Function<Enum,TypeResolution>` and the declaration can just pick up the arguments from the enclosing context inside the lambda.

Reply:
Inlined the static method. 

About `NAME_TO_METHOD`: Lambda can only pick up the fields `method` and `methodParameter` if the `NAME_TO_METHOD` is not static, otherwise the it has no access to the instance fields. Making it non-static on the other hand feels a bit wasteful.